test/sig_utils.py

```diff
@@ -301,3 +312,365 @@ def reference_vector_argument(arg):
+        arg_names = [f'decltype({arg.name})' for arg in args]
+        self.type_str = f"stan::return_type_t<{','.join(arg_names)}>"
+        self.__is_reverse_mode = any(arg.is_reverse_mode() for arg in args)
+    
+    def is_reverse_mode(self):
+        return self.__is_reverse_mode
+    
+    def cpp(self):
+        return f"{self.type_str} {self.name} = 0;"
+
+class RngArgument(CppStatement):
+    def __init__(self, name):
+        self.name = name
+    
+    def cpp(self):
+        return f"std::minstd_rand {self.name};"
+
+class OStreamArgument(CppStatement):
+    def __init__(self, name):
+        self.name = name
+    
+    def cpp(self):
+        return f"std::ostream* {self.name} = &std::cout;"
+
+class ArrayArgument(CppStatement):
+    def __init__(self, overload, name, number_nested_arrays, inner_type, value = None, size = None):
+        self.overload = overload
+        self.name = name
+        self.number_nested_arrays = number_nested_arrays
+        self.inner_type = inner_type
+
+        if not value:
+            if inner_type == "int":
+                self.value = 1
+            else:
+                self.value = 0.4
+        else:
+            self.value = value
+
+        if size:
+            self.size = size
+        else:
+            self.size = 1
+    
+    def is_reverse_mode(self):
+        return self.overload.startswith("Rev")
+    
+    def cpp(self):
+        scalar = overload_scalar[self.overload]
+        arg_type = arg_types[self.inner_type]
+        for n in range(self.number_nested_arrays):
+            arg_type = "std::vector<{}>".format(arg_type)
+        arg_type = arg_type.replace("SCALAR", scalar)
+
+        if isinstance(self.value, numbers.Number):
+            value_string = f"{self.value}"
+        else:
+            value_string = f"{{{','.join([repr(v) for v in self.value])}}}"
+
+        if isinstance(self.value, numbers.Number):
+            return f"auto {self.name} = stan::test::make_arg<{arg_type}>({self.value}, {self.size});"
+        else:
+            value_string = repr(self.value).replace('[', '{').replace(']', '}').replace('(', '{').replace(')', '}')
+            return f"{arg_type} {self.name} = {value_string};"
+
+class FunctionCallAssign(CppStatement):
+    def __init__(self, function_name, name, *args):
+        self.function_name = function_name
+        self.name = name
+        self.arg_str = ','.join(arg.name for arg in args)
+        # This isn't always accurate cause the output could be an int or a boolean
+        # or something even if the input is reverse mode
+        self.__is_reverse_mode = any(arg.is_reverse_mode() for arg in args)
+
+    def is_reverse_mode(self):
+        return self.__is_reverse_mode
+
+    def is_matrix_like(self):
+        raise Exception("is_matrix_like not implemented for FunctionCallAssign")
+
+    def cpp(self):
+        return f"auto {self.name} = stan::math::eval({self.function_name}({self.arg_str}));"
+
+class FunctionCall(CppStatement):
```

Comment:
I think `FunctionCall` and `FunctionCallAssignment` should be the same class, where `name` could be None.

Reply:
`FunctionCall(...)` and `FunctionCall(None, ...)` seems less descriptive than `FunctionCall(...)` and `FunctionCallAssign(...)` but anyway let me dig maybe these merge.